### PR TITLE
feat: Update remaining code to use new diagnostics

### PR DESCRIPTION
**Pull request by mark-koch**

### CODE REVIEW

File: guppylang/decorator.py [hunk 1]
```diff
@@ -149,7 +149,7 @@ def _get_python_caller(self, fn: PyFunc | None = None) -> ModuleIdentifier:
                         break
                 frame = frame.f_back
             else:
-                raise GuppyError("Could not find a caller for the `@guppy` decorator")
+                raise RuntimeError("Could not find a caller for the `@guppy` decorator")
```
Comment (mark-koch) [inline guppylang/decorator.py:152]: Guppy errors are now reserved for things that go wrong during compilation. This error is thrown immediately when the user calls the decorator, so we shouldn't use `GuppyError`

Same for other errors in this file

File: guppylang/decorator.py [hunk 2]
```diff
@@ -310,11 +310,6 @@ def custom(
 
         def dec(f: PyFunc) -> RawCustomFunctionDef:
             func_ast, docstring = parse_py_func(f, self._sources)
-            if not has_empty_body(func_ast):
```
Comment (croyzor) [inline guppylang/decorator.py:313]: What's going on here?

File: guppylang/experimental.py [hunk 3]
```diff
@@ -55,19 +58,30 @@ def __exit__(
...
 
 
+@dataclass(frozen=True)
+class ExperimentalFeatureError(Error):
+    title: ClassVar[str] = "Experimental feature"
+    span_label: ClassVar[str] = "{things} are an experimental feature"
+    things: str
+
+    @dataclass(frozen=True)
+    class Suggestion(Help):
```
Comment (croyzor) [inline guppylang/experimental.py:68]: Is this error ever going to be created without adding this Help?

File: guppylang/prelude/_internal/checker.py [hunk 4]
```diff
@@ -75,37 +79,19 @@ def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
...
     """Call checker for Python builtin functions that are not available in Guppy.
 
     Gives the uses a nicer error message when they try to use an unsupported feature.
     """
 
     def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
-        raise GuppyError(
-            f"Builtin method `{self.func.name}` is not supported by Guppy", self.node
-        )
+        err = UnsupportedError(self.node, f"Builtin method `{self.func.name}`")
```
Comment (croyzor) [inline guppylang/prelude/_internal/checker.py:89]: I think this should be `singular=True`

**Review by croyzor — APPROVED** [review]
New errors are looking very nice!

I'm seeing a lot of a pattern where an error declares a subdiagnostic as a nested class, which is added by the caller. Can we not define classes with default subdiagnostics?

**Comment by mark-koch** [comment]
> I'm seeing a lot of a pattern where an error declares a subdiagnostic as a nested class, which is added by the caller. Can we not define classes with default subdiagnostics?

I like the idea 👍  I created issue #620 to implement this

**Pull request merged**